### PR TITLE
fix(pairwise): normalize elig_mask value type at ingestion (#155, #158)

### PR DESCRIPTION
**Pull request by dgenio**
## 📋 Description

Pairwise eligibility consumers across `core`/`pairwise` only special-cased `(list, tuple)` (or `(list, tuple, np.ndarray)`) and silently treated **any other container** as "every operator eligible". Two *permitted* inputs hit that silent fallback:

- **#155** — a `set` mask, which `validate_pairwise_inputs` explicitly blesses, produced **incorrect, less-restrictive** eligibility (shifting `V_hat`/`SE`/`ESS`/`match_rate`, ~7% off in the audit).
- **#158** — the `np.ndarray` cells that `coerce_to_pandas` yields for Polars/PyArrow inputs broke the pandas-equivalence promised by #72 on the **pairwise** path (the sklearn path was already equivalent).

Both are the **same root cause** — type-sensitive `elig_mask` consumption — so both issues independently recommended the **same single-point fix**: normalize the value type once at ingestion rather than patching ~8 call sites (a partial patch is worse than none).

`PairwiseDesign.from_dataframes` now canonicalizes each `elig_mask` cell to a Python `list` once: `ndarray` → `tolist()`, `set`/`frozenset` → sorted, `tuple` → `list`; `NaN`/`None`/scalars are left untouched so the existing missing-mask "all eligible" fallback still applies. Every downstream consumer already handles `list`, so all paths (`standard` + `large_data`, propensity / candidate-pair / pre-split) become correct without touching the statistical call sites.

### Type of Change
- [x] 🐛 Bug fix (non-breaking change which fixes an issue)

## 🔗 Related Issues
- Fixes #155
- Fixes #158
- Related to #72

## 🧪 Testing

**Test commands run:**
```bash
ruff check src/ tests/ examples/        # All checks passed!
ruff format --check ...                  # 2 files already formatted
mypy src/skdr_eval/                      # Success: no issues found in 43 source files
python -m pytest tests/test_frame_inputs.py tests/test_pairwise_api.py   # 52 passed
python -m pytest                         # 895 passed, 9 skipped (full suite)
```

- [x] New and existing unit tests pass locally with my changes
- [x] Tested edge cases (`set`/`ndarray`/`tuple` cells, all-eligible baseline)

### Verified the tests catch the bug
With the normalization line disabled, the three new assertions fail (`test_pairwise_polars_inputs`, `test_set_mask_matches_list_mask[standard]`, `test_set_mask_matches_list_mask[large_data]`) with V_hat diverging 100%; they pass with the fix in place.

### Code Changes
- `src/skdr_eval/pairwise.py` — add `_normalize_elig_cell` helper; apply it to the `elig_col` column in `PairwiseDesign.from_dataframes` (the sole ingestion route; downstream of `coerce_to_pandas`).
- `tests/test_frame_inputs.py` — upgrade `test_pairwise_polars_inputs` from non-emptiness to exact pandas/Polars `V_hat` equivalence (#158); add `TestPairwiseEligMaskTypeEquivalence` asserting `set` ≡ `list` in `standard` and `large_data` modes and that a restrictive mask actually changes `V_hat` vs all-eligible (#155); refresh the stale #158 caveat.
- `CHANGELOG.md` — `### Fixed` entry.

### API Changes
- [x] No API changes (behavioral bug fix only; list-mask inputs are unaffected).

## 📚 Documentation
- [x] I have updated the CHANGELOG.md
- [x] Docstrings added for the new helper. README documents only the `elig_col` *name* (unchanged); `examples/quickstart_pairwise.py` uses list masks (unaffected), so no example/README change is warranted.

## ✅ Checklist
- [x] `ruff check` and `ruff format` clean
- [x] `mypy` clean
- [x] Full suite green (895 passed, 9 skipped)
- [x] New tests fail without the fix, pass with it

### Focus Areas
- **Order sensitivity:** `set` → `sorted()` reorders elements vs the original list, but eligibility is membership-based — pair-construction order derives from `day_ops`, not from `elig_mask` order — so `V_hat` is invariant to mask order. The `set ≡ list` equivalence test confirms this empirically.
- **Statistical-logic change:** this alters eligibility (a `V_hat` input). The equivalence tests act as the proof: the already-validated `list` path is ground truth, and the fix makes `set`/`ndarray` paths reproduce it exactly while a restrictive mask provably differs from all-eligible.

https://claude.ai/code/session_01WajZY9ttk9RPs5kxyVNvaA

---
_Generated by [Claude Code](https://claude.ai/code/session_01WajZY9ttk9RPs5kxyVNvaA)_

### CODE REVIEW

File: src/skdr_eval/pairwise.py [hunk 1]
```diff
@@ -16,6 +16,41 @@
...
+    ``None``, scalars) are returned unchanged so the existing
+    "missing mask ⇒ all eligible" fallback still applies.
+    """
+    if isinstance(value, np.ndarray):
+        return value.tolist()
+    if isinstance(value, (set, frozenset)):
+        try:
+            return sorted(value)
+        except TypeError:
+            return list(value)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/pairwise.py]: Good catch — fixed in 30c0a8f. The `TypeError` branch in `_normalize_elig_cell` (`src/skdr_eval/pairwise.py:42-57`) now returns `sorted(value, key=repr)` instead of `list(value)`, so mixed/unorderable element types get a stable order across runs and Python versions and the stored design stays reproducible. The docstring was updated to explain the `repr`-keyed fallback, and `TestNormalizeEligCell.test_unorderable_set_uses_repr_keyed_fallback` covers it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_frame_inputs.py [hunk 2]
```diff
@@ -92,26 +117,75 @@ def test_sklearn_arrow_matches_pandas(self) -> None:
...
+
+class TestPairwiseEligMaskTypeEquivalence:
+    """A set-valued elig_mask must be honored identically to a list (#155).
+
+    ``validate_pairwise_inputs`` blesses ``set`` masks, but most eligibility
+    consumers only special-case ``(list, tuple)`` and silently fell back to
+    "every operator eligible" for a ``set`` — producing incorrect, less
+    restrictive eligibility. Normalizing at ingestion makes a ``set`` mask
+    produce the same V_hat as the identical mask expressed as a ``list``.
+    """
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_frame_inputs.py:221]: Agreed — this is exactly the repo's statistical-integrity requirement, and it's addressed in aff96f9. New `tests/sim_studies/test_elig_mask_recovery.py` builds an eligibility-sensitive pairwise DGP with an analytic constrained optimum `V*_restricted` and asserts the induced-policy DR estimate recovers it, while the unconstrained `V*_full` (the pre-fix bug attractor) is strictly better. It then asserts `list`/`set`/`ndarray` masks all recover the same `V*` and that "closer to V*_restricted than to V*_full" holds — verified to fail when the normalization is removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_frame_inputs.py [hunk 3]
```diff
@@ -92,26 +117,75 @@ def test_sklearn_arrow_matches_pandas(self) -> None:
...
+        art_list = _evaluate_pairwise(
+            logs_df, op_daily_df, execution_mode=execution_mode
+        )
+        art_set = _evaluate_pairwise(
+            _with_set_masks(logs_df), op_daily_df, execution_mode=execution_mode
+        )
+        pd.testing.assert_series_equal(
+            art_list.report["V_hat"].reset_index(drop=True),
+            art_set.report["V_hat"].reset_index(drop=True),
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_frame_inputs.py]: Done in 30c0a8f. The equivalence checks now compare `_ELIG_SENSITIVE_COLS = ["V_hat", "SE_if", "ESS", "match_rate"]` via the `_assert_pairwise_report_equiv` helper (`tests/test_frame_inputs.py:65-73`), so `TestPairwiseEligMaskTypeEquivalence.test_set_mask_matches_list_mask` now fails on any support-diagnostic divergence, not just `V_hat`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_frame_inputs.py [hunk 4]
```diff
@@ -92,26 +117,75 @@ def test_sklearn_arrow_matches_pandas(self) -> None:
...
-        n_splits=3,
-        fit_models=True,
-        policy_train="pre_split",
-        random_state=0,
+    art_pd = _evaluate_pairwise(logs_df, op_daily_df)
+    art_pl = _evaluate_pairwise(pl.from_pandas(logs_df), pl.from_pandas(op_daily_df))
+    pd.testing.assert_series_equal(
+        art_pd.report["V_hat"].reset_index(drop=True),
+        art_pl.report["V_hat"].reset_index(drop=True),
     )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_frame_inputs.py]: Done in 30c0a8f — same fix as the `set ≡ list` check. `test_pairwise_polars_inputs` now calls `_assert_pairwise_report_equiv`, comparing the full eligibility-sensitive column set (`V_hat`, `SE_if`, `ESS`, `match_rate`) between the pandas and Polars paths instead of asserting on `V_hat` alone.

---
_Generated by [Claude Code](https://claude.ai/code)_
